File: 64.Hopcraft-KarpAlgo.py
# The Hopcroft-Karp Algorithm is an efficient algorithm to find the maximum matching in a bipartite graph. 
# It improves over the simple augmenting path approach
# by finding multiple augmenting paths simultaneously, achieving a time complexity of O(√V * E).

# Explanation

# A bipartite graph consists of two disjoint sets of vertices, say U and V, 
# where edges only connect vertices between sets, not within a set.

# A matching is a set of edges such that no two edges share a vertex.

# The maximum matching is a matching with the largest possible number of edges.

# The Hopcroft-Karp algorithm uses a BFS + DFS strategy:

# BFS (Breadth First Search) to layer the graph and find all shortest augmenting paths.

# DFS (Depth First Search) to augment along these paths in parallel.

# By processing multiple paths at once, it reduces redundant searches, 
# making it much faster than a naive augmenting-path algorithm.

from collections import deque

class HopcroftKarp:
    def __init__(self, U, V):

        self.U = U
        self.V = V
        self.pair_U = [-1] * U  # Stores matched vertex in V for U
        self.pair_V = [-1] * V  # Stores matched vertex in U for V
        self.dist = [0] * U     # Distance labels for BFS
        self.graph = [[] for _ in range(U)]  # adjacency list for U

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def bfs(self):
        queue = deque()
        for u in range(self.U):
            if self.pair_U[u] == -1:
                self.dist[u] = 0
                queue.append(u)
            else:
                self.dist[u] = float('inf')

        found = False
        while queue:
            u = queue.popleft()
            for v in self.graph[u]:
                if self.pair_V[v] == -1:
                    found = True  # Found an augmenting path
                elif self.dist[self.pair_V[v]] == float('inf'):
                    self.dist[self.pair_V[v]] = self.dist[u] + 1
                    queue.append(self.pair_V[v])
        return found

    def dfs(self, u):
        for v in self.graph[u]:
            if self.pair_V[v] == -1 or (self.dist[self.pair_V[v]] == self.dist[u] + 1 and self.dfs(self.pair_V[v])):
                self.pair_U[u] = v
                self.pair_V[v] = u
                return True
        self.dist[u] = float('inf')
        return False

    def max_matching(self):
        result = 0
        while self.bfs():
            for u in range(self.U):
                if self.pair_U[u] == -1:
                    if self.dfs(u):
                        result += 1
        return result


hk = HopcroftKarp(4, 4)
hk.add_edge(0, 0)
hk.add_edge(0, 1)
hk.add_edge(1, 1)
hk.add_edge(2, 2)
hk.add_edge(3, 2)
hk.add_edge(3, 3)

print("Maximum Matching Size:", hk.max_matching())
print("Pairs U -> V:", hk.pair_U)
print("Pairs V -> U:", hk.pair_V)

File: 64.Hopcraft-KarpAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio      */
/* -----------------------------------------------------------------------  */

import java.util.*;

public class HopcroftKarp {
    int U, V;
    int[] pairU, pairV, dist;
    List<List<Integer>> graph;

    // Constructor
    public HopcroftKarp(int U, int V) {
        this.U = U;
        this.V = V;
        pairU = new int[U];
        pairV = new int[V];
        dist = new int[U];

        Arrays.fill(pairU, -1);
        Arrays.fill(pairV, -1);

        graph = new ArrayList<>();
        for (int i = 0; i < U; i++) {
            graph.add(new ArrayList<>());
        }
    }

    // Add edge from u in U to v in V
    public void addEdge(int u, int v) {
        graph.get(u).add(v);
    }

    // BFS to build layers
    private boolean bfs() {
        Queue<Integer> queue = new LinkedList<>();
        for (int u = 0; u < U; u++) {
            if (pairU[u] == -1) {
                dist[u] = 0;
                queue.add(u);
            } else {
                dist[u] = Integer.MAX_VALUE;
            }
        }

        boolean found = false;
        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (int v : graph.get(u)) {
                if (pairV[v] == -1) {
                    found = true; // Found an augmenting path
                } else if (dist[pairV[v]] == Integer.MAX_VALUE) {
                    dist[pairV[v]] = dist[u] + 1;
                    queue.add(pairV[v]);
                }
            }
        }
        return found;
    }

    // DFS to find augmenting paths
    private boolean dfs(int u) {
        for (int v : graph.get(u)) {
            if (pairV[v] == -1 || (dist[pairV[v]] == dist[u] + 1 && dfs(pairV[v]))) {
                pairU[u] = v;
                pairV[v] = u;
                return true;
            }
        }
        dist[u] = Integer.MAX_VALUE;
        return false;
    }

    // Main method to compute maximum matching
    public int maxMatching() {
        int result = 0;
        while (bfs()) {
            for (int u = 0; u < U; u++) {
                if (pairU[u] == -1) {
                    if (dfs(u)) {
                        result++;
                    }
                }
            }
        }
        return result;
    }

    // Example usage
    public static void main(String[] args) {
        HopcroftKarp hk = new HopcroftKarp(4, 4);
        hk.addEdge(0, 0);
        hk.addEdge(0, 1);
        hk.addEdge(1, 1);
        hk.addEdge(2, 2);
        hk.addEdge(3, 2);
        hk.addEdge(3, 3);

        System.out.println("Maximum Matching Size: " + hk.maxMatching());
        System.out.println("Pairs U -> V: " + Arrays.toString(hk.pairU));
        System.out.println("Pairs V -> U: " + Arrays.toString(hk.pairV));
    }
}
